README
README - 2011-11-19
-------------------


INTRODUCTION

    This README file describes the microxml library version 1.0.

    microxml is a small XML parsing library that you can use to read XML and
    XML-like data files in your application without requiring large non-standard
    libraries.  microxml only requires an ANSI C compatible compiler (GCC works,
    as do most vendors' ANSI C compilers) and a "make" program.

    microxml provides the following functionality:

	- Reading of UTF-8 and UTF-16 and writing of UTF-8 encoded XML files and
	  strings.
	- Data is stored in a linked-list tree structure, preserving the XML
	  data hierarchy.
	- Supports arbitrary element names, attributes, and attribute values
	  with no preset limits, just available memory.
	- Supports integer, real, opaque ("cdata"), and text data types in
	  "leaf" nodes.
	- Functions for creating and managing trees of data.
	- "Find" and "walk" functions for easily locating and navigating trees
	  of data.

    microxml doesn't do validation or other types of processing on the data
    based upon schema files or other sources of definition information.


BUILDING microxml

    microxml comes with an autoconf-based configure script; just type the
    following command to get things going:

        ./configure

    The default install prefix is /usr/local, which can be overridden using the
    --prefix option:

        ./configure --prefix=/foo

    Other configure options can be found using the --help option:

        ./configure --help

    Once you have configured the software, type "make" to do the build and run
    the test program to verify that things are working, as follows:

        make

    If you are using microxml under Microsoft Windows with Visual C++ 2008, use
    the included project files in the "vcnet" subdirectory to build the library
    instead.


INSTALLING microxml

    The "install" target will install microxml in the lib and include
    directories:

        make install

    Once you have installed it, use the "-lmicroxml" option to link your application
    against it.


    microxml provides a single header file which you include:

        #include <microxml.h>

    Nodes are defined by the "mxml_node_t" structure; the "type" member defines
    the node type (element, integer, opaque, real, or text) which determines
    which value you want to look at in the "value" union.  New nodes can be
    created using the "mxmlNewElement()", "mxmlNewInteger()", "mxmlNewOpaque()",
    "mxmlNewReal()", and "mxmlNewText()" functions.  Only elements can have
    child nodes, and the top node must be an element, usually "?xml".

    You load an XML file using the "mxmlLoadFile()" function:

        FILE *fp;
        mxml_node_t *tree;

	fp = fopen("filename.xml", "r");
	tree = mxmlLoadFile(NULL, fp, MXML_NO_CALLBACK);
	fclose(fp);

    Similarly, you save an XML file using the "mxmlSaveFile()" function:

        FILE *fp;
        mxml_node_t *tree;

	fp = fopen("filename.xml", "w");
	mxmlSaveFile(tree, fp, MXML_NO_CALLBACK);
	fclose(fp);

    The "mxmlLoadString()", "mxmlSaveAllocString()", and "mxmlSaveString()"
    functions load XML node trees from and save XML node trees to strings:

        char buffer[8192];
	char *ptr;
	mxml_node_t *tree;

        ...
	tree = mxmlLoadString(NULL, buffer, MXML_NO_CALLBACK);

        ...
        mxmlSaveString(tree, buffer, sizeof(buffer), MXML_NO_CALLBACK);

        ...
	ptr = mxmlSaveAllocString(tree, MXML_NO_CALLBACK);

    You can find a named element/node using the "mxmlFindElement()" function:

        mxml_node_t *node = mxmlFindElement(tree, tree, "name", "attr",
	                                    "value", MXML_DESCEND);

    The "name", "attr", and "value" arguments can be passed as NULL to act as
    wildcards, e.g.:

        /* Find the first "a" element */
        node = mxmlFindElement(tree, tree, "a", NULL, NULL, MXML_DESCEND);

        /* Find the first "a" element with "href" attribute */
        node = mxmlFindElement(tree, tree, "a", "href", NULL, MXML_DESCEND);

        /* Find the first "a" element with "href" to a URL */
        node = mxmlFindElement(tree, tree, "a", "href",
	                       "http://www.minixml.org/",
			       MXML_DESCEND);

        /* Find the first element with a "src" attribute*/
        node = mxmlFindElement(tree, tree, NULL, "src", NULL, MXML_DESCEND);

        /* Find the first element with a "src" = "foo.jpg" */
        node = mxmlFindElement(tree, tree, NULL, "src", "foo.jpg",
	                       MXML_DESCEND);

    You can also iterate with the same function:

        mxml_node_t *node;

	for (node = mxmlFindElement(tree, tree, "name", NULL, NULL,
	                            MXML_DESCEND);
	     node != NULL;
	     node = mxmlFindElement(node, tree, "name", NULL, NULL,
	                            MXML_DESCEND))
        {
	  ... do something ...
	}

    The "mxmlFindPath()" function finds the (first) value node under a specific
    element using a "path":

        mxml_node_t *value = mxmlFindPath(tree, "path/to/*/foo/bar");

    The "mxmlGetInteger()", "mxmlGetOpaque()", "mxmlGetReal()", and
    "mxmlGetText()" functions retrieve the value from a node:

        mxml_node_t *node;

        int intvalue = mxmlGetInteger(node);

        const char *opaquevalue = mxmlGetOpaque(node);

        double realvalue = mxmlGetReal(node);

        int whitespacevalue;
        const char *textvalue = mxmlGetText(node, &whitespacevalue);

    Finally, once you are done with the XML data, use the "mxmlDelete()"
    function to recursively free the memory that is used for a particular node
    or the entire tree:

        mxmlDelete(tree);


GETTING HELP AND REPORTING PROBLEMS

    TODO

LEGAL STUFF

    The Mini-XML library is Copyright 2003-2011 by Michael Sweet.  License terms
    are described in the file "COPYING".

    The microxml library is Copyright 2011-2012 by Luka Perkov.  License terms
    are described in the file "COPYING".



CHANGES
CHANGES - 2012-01-07
--------------------

CHANGES IN microxml 1.0

	- microxml becomes a fork of Mini-XML (r446)
	- Renamed relevant parts of mxml to microxml
	- Added functions


CHANGES IN Mini-XML 2.7

	- Added 64-bit configurations to the VC++ project files (STR #129)
	- Fixed conformance of mxmldoc's HTML and CSS output.
	- Added data accessor ("get") functions and made the mxml_node_t and
	  mxml_index_t structures private but still available in the Mini-XML
	  header to preserve source compatibility (STR #118)
	- Updated the source headers to reference the Mini-XML license and its
	  exceptions to the LGPL2 (STR #108)
	- Fixed a memory leak when loading a badly-formed XML file (STR #121)
	- Added a new mxmlFindPath() function to find the value node of a
	  named element (STR #110)
	- Building a static version of the library did not work on Windows
	  (STR #112)
	- The shared library did not include a destructor for the thread-
	  specific data key on UNIX-based operating systems (STR #103)
	- mxmlLoad* did not error out on XML with multiple root nodes (STR #101)
	- Fixed an issue with the _mxml_vstrdupf function (STR #107)
	- mxmlSave* no longer write all siblings of the passed node, just that
	  node and its children (STR #109)


CHANGES IN Mini-XML 2.6

	- Documentation fixes (STR #91, STR #92)
	- The mxmldoc program did not handle typedef comments properly (STR #72)
	- Added support for "long long" printf formats.
	- The XML parser now ignores BOMs in UTF-8 XML files (STR #89)
	- The mxmldoc program now supports generating Xcode documentation sets.
	- mxmlSave*() did not output UTF-8 correctly on some platforms.
	- mxmlNewXML() now adds encoding="utf-8" in the ?xml directive to avoid
	  problems with non-conformant XML parsers that assume something other
	  than UTF-8 as the default encoding.
	- Wrapping was not disabled when mxmlSetWrapMargin(0) was called, and
	  "<?xml ... ?>" was always followed by a newline (STR #76)
	- The mxml.pc.in file was broken (STR #79)
	- The mxmldoc program now handles "typedef enum name {} name" correctly
	  (STR #72)


CHANGES IN Mini-XML 2.5

	- The mxmldoc program now makes greater use of CSS and
	  supports a --css option to embed an alternate stylesheet.
	- The mxmldoc program now supports --header and --footer
	  options to insert documentation content before and
	  after the generated content.
	- The mxmldoc program now supports a --framed option to
	  generate framed HTML output.
	- The mxmldoc program now creates a table of contents
	  including any headings in the --intro file when
	  generating HTML output.
	- The man pages and man page output from mxmldoc did
	  not use "\-" for dashes (STR #68)
	- The debug version of the Mini-XML DLL could not be
	  built (STR #65)
	- Processing instructions and directives did not work
	  when not at the top level of a document (STR #67)
	- Spaces around the "=" in attributes were not supported
	  (STR #67)


CHANGES IN Mini-XML 2.4

	- Fixed shared library build problems on HP-UX and Mac OS X.
	- The mxmldoc program did not output argument descriptions
	  for functions properly.
	- All global settings (custom, error, and entity callbacks
	  and the wrap margin) are now managed separately for each
	  thread.
	- Added mxmlElementDeleteAttr() function (STR #59)
	- mxmlElementSetAttrf() did not work (STR #57)
	- mxmlLoad*() incorrectly treated declarations as parent
	  elements (STR #56)
	- mxmlLoad*() incorrectly allowed attributes without values
	  (STR #47)
	- Fixed Visual C++ build problems (STR #49)
	- mxmlLoad*() did not return NULL when an element contained
	  an error (STR #46)
	- Added support for the apos character entity (STR #54)
	- Fixed whitespace detection with Unicode characters (STR
	  #48)
	- mxmlWalkNext() and mxmlWalkPrev() did not work correctly
	  when called with a node with no children as the top node
	  (STR #53)


CHANGES IN Mini-XML 2.3

	- Added two exceptions to the LGPL to support static
	  linking of applications against Mini-XML
	- The mxmldoc utility can now generate man pages, too.
	- Added a mxmlNewXML() function
	- Added a mxmlElementSetAttrf() function (STR #43)
	- Added snprintf() emulation function for test program (STR
	  #32)
	- Added the _CRT_SECURE_NO_DEPRECATE definition when
	  building on VC++ 2005 (STR #36)
	- mxmlLoad*() did not detect missing > characters in
	  elements (STR #41)
	- mxmlLoad*() did not detect missing close tags at the end
	  of an XML document (STR #45)
	- Added user_data and ref_count members to mxml_node_t
	  structure
	- Added mxmlReleaseNode() and mxmlRetainNode() APIs for
	  reference-counted nodes
	- Added mxmlSetWrapMargin() to control the wrapping of XML
	  output
	- Added conditional check for EINTR error code for
	  certain Windows compilers that do not define it (STR
	  #33)
	- The mxmldoc program now generates correct HTML 4.0
	  output - previously it generated invalid XHTML
	- The mxmldoc program now supports "@deprecated@,
	  "@private@", and "@since version@" comments
	- Fixed function and enumeration type bugs in mxmldoc.
	- Fixed the XML schema for mxmldoc
	- The mxmldoc program now supports --intro, --section,
	  and --title options
	- The mxmlLoad*() functions could leak a node on an error
	  (STR #27)
	- The mxml_vsnprintf() function could get in an infinite
	  loop on a buffer overflow (STR #25)
	- Added new mxmlNewCDATA() and mxmlSetCDATA() functions
	  to create and set CDATA nodes, which are really just
	  special element nodes
	- Added new MXML_IGNORE type and MXML_IGNORE_CB callback
	  to ignore non-element nodes, e.g. whitespace
	- mxmlLoad*() crashed when reporting an error in some
	  invalid XML (STR #23)


CHANGES IN Mini-XML 2.2.2


mxml-node.c
/*
 * Node support code for microxml, a micro XML-like file parsing library.
 *
 * Copyright 2003-2011 by Michael R Sweet.
 * Copyright 2011-2012 by Luka Perkov.
 *
 * These coded instructions, statements, and computer programs are the
 * property of Michael R Sweet and Luka Perkov. They are protected by
 * Federal copyright law. Distribution and use rights are outlined in
 * the file "COPYING" which should have been included with this file.
 * If this file is missing or damaged, see the license at:
 *
 *     http://www.minixml.org/
 *
 * Contents:
 *
 *   mxmlAdd()         - Add a node to a tree.
 *   mxmlDelete()      - Delete a node and all of its children.
 *   mxmlGetRefCount() - Get the current reference (use) count for a node.
 *   mxmlNewCDATA()    - Create a new CDATA node.
 *   mxmlNewCustom()   - Create a new custom data node.
 *   mxmlNewElement()  - Create a new element node.
 *   mxmlNewInteger()  - Create a new integer node.
 *   mxmlNewOpaque()   - Create a new opaque string.
 *   mxmlNewReal()     - Create a new real number node.
 *   mxmlNewText()     - Create a new text fragment node.
 *   mxmlNewTextf()    - Create a new formatted text fragment node.
 *   mxmlRemove()      - Remove a node from its parent.
 *   mxmlNewXML()      - Create a new XML document tree.
 *   mxmlRelease()     - Release a node.
 *   mxmlRetain()      - Retain a node.
 *   mxml_new()        - Create a new node.
 */

#include "config.h"
#include "microxml.h"


/*
 * Local functions...
 */

static mxml_node_t	*mxml_new(mxml_node_t *parent, mxml_type_t type);


/*
 * 'mxmlAdd()' - Add a node to a tree.
 *
 * Adds the specified node to the parent. If the child argument is not
 * NULL, puts the new node before or after the specified child depending
 * on the value of the where argument. If the child argument is NULL,
 * puts the new node at the beginning of the child list (MXML_ADD_BEFORE)
 * or at the end of the child list (MXML_ADD_AFTER). The constant
 * MXML_ADD_TO_PARENT can be used to specify a NULL child pointer.
 */

void
mxmlAdd(mxml_node_t *parent,		/* I - Parent node */
        int         where,		/* I - Where to add, MXML_ADD_BEFORE or MXML_ADD_AFTER */
        mxml_node_t *child,		/* I - Child node for where or MXML_ADD_TO_PARENT */
	mxml_node_t *node)		/* I - Node to add */
{
#ifdef DEBUG
  fprintf(stderr, "mxmlAdd(parent=%p, where=%d, child=%p, node=%p)\n", parent,
          where, child, node);
#endif /* DEBUG */

 /*
  * Range check input...
  */

  if (!parent || !node)
    return;

#if DEBUG > 1
  fprintf(stderr, "    BEFORE: node->parent=%p\n", node->parent);
  if (parent)
  {
    fprintf(stderr, "    BEFORE: parent->child=%p\n", parent->child);
    fprintf(stderr, "    BEFORE: parent->last_child=%p\n", parent->last_child);
    fprintf(stderr, "    BEFORE: parent->prev=%p\n", parent->prev);
    fprintf(stderr, "    BEFORE: parent->next=%p\n", parent->next);
  }
#endif /* DEBUG > 1 */

 /*
  * Remove the node from any existing parent...
  */

  if (node->parent)
    mxmlRemove(node);

 /*
  * Reset pointers...
  */

  node->parent = parent;

  switch (where)
  {
    case MXML_ADD_BEFORE :
        if (!child || child == parent->child || child->parent != parent)
	{
	 /*
	  * Insert as first node under parent...
	  */

	  node->next = parent->child;

	  if (parent->child)
	    parent->child->prev = node;
	  else
	    parent->last_child = node;

	  parent->child = node;
	}
	else
	{
	 /*
	  * Insert node before this child...
	  */

	  node->next = child;
	  node->prev = child->prev;

	  if (child->prev)
	    child->prev->next = node;
	  else
	    parent->child = node;

	  child->prev = node;
	}
        break;

    case MXML_ADD_AFTER :
        if (!child || child == parent->last_child || child->parent != parent)
	{
	 /*
	  * Insert as last node under parent...
	  */

	  node->parent = parent;
	  node->prev   = parent->last_child;

	  if (parent->last_child)
	    parent->last_child->next = node;
	  else
	    parent->child = node;

	  parent->last_child = node;
        }
	else
	{
	 /*
	  * Insert node after this child...
	  */

	  node->prev = child;
	  node->next = child->next;

	  if (child->next)
	    child->next->prev = node;
	  else
	    parent->last_child = node;

	  child->next = node;
	}
        break;
  }

#if DEBUG > 1
  fprintf(stderr, "    AFTER: node->parent=%p\n", node->parent);
  if (parent)
  {
    fprintf(stderr, "    AFTER: parent->child=%p\n", parent->child);
    fprintf(stderr, "    AFTER: parent->last_child=%p\n", parent->last_child);
    fprintf(stderr, "    AFTER: parent->prev=%p\n", parent->prev);
    fprintf(stderr, "    AFTER: parent->next=%p\n", parent->next);
  }
#endif /* DEBUG > 1 */
}


/*
 * 'mxmlDelete()' - Delete a node and all of its children.
 *
 * If the specified node has a parent, this function first removes the
 * node from its parent using the mxmlRemove() function.
 */

void
mxmlDelete(mxml_node_t *node)		/* I - Node to delete */
{
  int	i;				/* Looping var */


#ifdef DEBUG
  fprintf(stderr, "mxmlDelete(node=%p)\n", node);
#endif /* DEBUG */

 /*
  * Range check input...
  */

  if (!node)
    return;

 /*
  * Remove the node from its parent, if any...
  */

  mxmlRemove(node);

 /*
  * Delete children...
  */

  while (node->child)
    mxmlDelete(node->child);

 /*
  * Now delete any node data...
  */

  switch (node->type)
  {
    case MXML_ELEMENT :
        if (node->value.element.name)
          free(node->value.element.name);

        if (node->value.element.num_attrs)
        {
          for (i = 0; i < node->value.element.num_attrs; i ++)
          {
            if (node->value.element.attrs[i].name)
              free(node->value.element.attrs[i].name);
            if (node->value.element.attrs[i].value)
              free(node->value.element.attrs[i].value);
          }

          free(node->value.element.attrs);
        }
        break;
    case MXML_INTEGER :
       /* Nothing to do */
        break;
    case MXML_OPAQUE :
        if (node->value.opaque)
	  free(node->value.opaque);
        break;
    case MXML_REAL :
       /* Nothing to do */
        break;
    case MXML_TEXT :
        if (node->value.text.string)
	  free(node->value.text.string);
        break;
    case MXML_CUSTOM :
        if (node->value.custom.data &&
	    node->value.custom.destroy)
	  (*(node->value.custom.destroy))(node->value.custom.data);
	break;
    default :
        break;
  }

 /*
  * Free this node...
  */

  free(node);
}


/*
 * 'mxmlGetRefCount()' - Get the current reference (use) count for a node.
 *
 * The initial reference count of new nodes is 1. Use the @link mxmlRetain@
 * and @link mxmlRelease@ functions to increment and decrement a node's
 * reference count.
 *
 * @since Mini-XML 2.7@.
 */

int					/* O - Reference count */
mxmlGetRefCount(mxml_node_t *node)	/* I - Node */
{
 /*
  * Range check input...
  */

  if (!node)
    return (0);

 /*
  * Return the reference count...
  */

  return (node->ref_count);
}


/*
 * 'mxmlNewCDATA()' - Create a new CDATA node.
 *
 * The new CDATA node is added to the end of the specified parent's child
 * list. The constant MXML_NO_PARENT can be used to specify that the new
 * CDATA node has no parent. The data string must be nul-terminated and
 * is copied into the new node. CDATA nodes use the MXML_ELEMENT type.
 *
 * @since Mini-XML 2.3@
 */

mxml_node_t *				/* O - New node */
mxmlNewCDATA(mxml_node_t *parent,	/* I - Parent node or MXML_NO_PARENT */
	     const char  *data)		/* I - Data string */
{
  mxml_node_t	*node;			/* New node */


#ifdef DEBUG
  fprintf(stderr, "mxmlNewCDATA(parent=%p, data=\"%s\")\n",
          parent, data ? data : "(null)");
#endif /* DEBUG */

 /*
  * Range check input...
  */

  if (!data)
    return (NULL);

 /*
  * Create the node and set the name value...
  */

  if ((node = mxml_new(parent, MXML_ELEMENT)) != NULL)
    node->value.element.name = _mxml_strdupf("![CDATA[%s]]", data);

  return (node);
}


/*
 * 'mxmlNewCustom()' - Create a new custom data node.
 *
 * The new custom node is added to the end of the specified parent's child
 * list. The constant MXML_NO_PARENT can be used to specify that the new
 * element node has no parent. NULL can be passed when the data in the
 * node is not dynamically allocated or is separately managed.
 *
 * @since Mini-XML 2.1@
 */

mxml_node_t *				/* O - New node */
mxmlNewCustom(
    mxml_node_t              *parent,	/* I - Parent node or MXML_NO_PARENT */
    void                     *data,	/* I - Pointer to data */
    mxml_custom_destroy_cb_t destroy)	/* I - Function to destroy data */
{
  mxml_node_t	*node;			/* New node */


#ifdef DEBUG
  fprintf(stderr, "mxmlNewCustom(parent=%p, data=%p, destroy=%p)\n", parent,
          data, destroy);
#endif /* DEBUG */

 /*
  * Create the node and set the value...
  */

  if ((node = mxml_new(parent, MXML_CUSTOM)) != NULL)
  {
    node->value.custom.data    = data;
    node->value.custom.destroy = destroy;
  }

  return (node);
}


/*
 * 'mxmlNewElement()' - Create a new element node.
 *
 * The new element node is added to the end of the specified parent's child
 * list. The constant MXML_NO_PARENT can be used to specify that the new
 * element node has no parent.
 */

mxml_node_t *				/* O - New node */
mxmlNewElement(mxml_node_t *parent,	/* I - Parent node or MXML_NO_PARENT */
               const char  *name)	/* I - Name of element */
{
  mxml_node_t	*node;			/* New node */


#ifdef DEBUG
  fprintf(stderr, "mxmlNewElement(parent=%p, name=\"%s\")\n", parent,
          name ? name : "(null)");
#endif /* DEBUG */

 /*
  * Range check input...
  */

  if (!name)
    return (NULL);

 /*
  * Create the node and set the element name...
  */

  if ((node = mxml_new(parent, MXML_ELEMENT)) != NULL)
    node->value.element.name = strdup(name);

  return (node);
}


/*
 * 'mxmlNewInteger()' - Create a new integer node.
 *
 * The new integer node is added to the end of the specified parent's child
 * list. The constant MXML_NO_PARENT can be used to specify that the new
 * integer node has no parent.
 */

mxml_node_t *				/* O - New node */
mxmlNewInteger(mxml_node_t *parent,	/* I - Parent node or MXML_NO_PARENT */
               int         integer)	/* I - Integer value */
{
  mxml_node_t	*node;			/* New node */


#ifdef DEBUG
  fprintf(stderr, "mxmlNewInteger(parent=%p, integer=%d)\n", parent, integer);
#endif /* DEBUG */

 /*
  * Create the node and set the element name...
  */

  if ((node = mxml_new(parent, MXML_INTEGER)) != NULL)
    node->value.integer = integer;

  return (node);
}


/*
 * 'mxmlNewOpaque()' - Create a new opaque string.
 *
 * The new opaque node is added to the end of the specified parent's child
 * list. The constant MXML_NO_PARENT can be used to specify that the new
 * opaque node has no parent. The opaque string must be nul-terminated and
 * is copied into the new node.
 */

mxml_node_t *				/* O - New node */
mxmlNewOpaque(mxml_node_t *parent,	/* I - Parent node or MXML_NO_PARENT */
              const char  *opaque)	/* I - Opaque string */
{
  mxml_node_t	*node;			/* New node */


#ifdef DEBUG
  fprintf(stderr, "mxmlNewOpaque(parent=%p, opaque=\"%s\")\n", parent,
          opaque ? opaque : "(null)");
#endif /* DEBUG */

 /*
  * Range check input...
  */

  if (!opaque)
    return (NULL);

 /*
  * Create the node and set the element name...
  */

  if ((node = mxml_new(parent, MXML_OPAQUE)) != NULL)
    node->value.opaque = strdup(opaque);

  return (node);
}


/*
 * 'mxmlNewReal()' - Create a new real number node.
 *
 * The new real number node is added to the end of the specified parent's
 * child list. The constant MXML_NO_PARENT can be used to specify that
 * the new real number node has no parent.
 */

mxml_node_t *				/* O - New node */
mxmlNewReal(mxml_node_t *parent,	/* I - Parent node or MXML_NO_PARENT */
            double      real)		/* I - Real number value */
{
  mxml_node_t	*node;			/* New node */


#ifdef DEBUG
  fprintf(stderr, "mxmlNewReal(parent=%p, real=%g)\n", parent, real);
#endif /* DEBUG */

 /*
  * Create the node and set the element name...
  */

  if ((node = mxml_new(parent, MXML_REAL)) != NULL)
    node->value.real = real;

  return (node);
}


/*
 * 'mxmlNewText()' - Create a new text fragment node.
 *
 * The new text node is added to the end of the specified parent's child
 * list. The constant MXML_NO_PARENT can be used to specify that the new
 * text node has no parent. The whitespace parameter is used to specify
 * whether leading whitespace is present before the node. The text
 * string must be nul-terminated and is copied into the new node.  
 */

mxml_node_t *				/* O - New node */
mxmlNewText(mxml_node_t *parent,	/* I - Parent node or MXML_NO_PARENT */
            int         whitespace,	/* I - 1 = leading whitespace, 0 = no whitespace */
	    const char  *string)	/* I - String */
{
  mxml_node_t	*node;			/* New node */


#ifdef DEBUG
  fprintf(stderr, "mxmlNewText(parent=%p, whitespace=%d, string=\"%s\")\n",
          parent, whitespace, string ? string : "(null)");
#endif /* DEBUG */

 /*
  * Range check input...
  */

  if (!string)
    return (NULL);

 /*
  * Create the node and set the text value...
  */

  if ((node = mxml_new(parent, MXML_TEXT)) != NULL)
  {
    node->value.text.whitespace = whitespace;
    node->value.text.string     = strdup(string);
  }

  return (node);
}


/*
 * 'mxmlNewTextf()' - Create a new formatted text fragment node.
 *
 * The new text node is added to the end of the specified parent's child
 * list. The constant MXML_NO_PARENT can be used to specify that the new
 * text node has no parent. The whitespace parameter is used to specify
 * whether leading whitespace is present before the node. The format
 * string must be nul-terminated and is formatted into the new node.  
 */

mxml_node_t *				/* O - New node */
mxmlNewTextf(mxml_node_t *parent,	/* I - Parent node or MXML_NO_PARENT */
             int         whitespace,	/* I - 1 = leading whitespace, 0 = no whitespace */
	     const char  *format,	/* I - Printf-style frmat string */
	     ...)			/* I - Additional args as needed */
{
  mxml_node_t	*node;			/* New node */
  va_list	ap;			/* Pointer to arguments */


#ifdef DEBUG
  fprintf(stderr, "mxmlNewTextf(parent=%p, whitespace=%d, format=\"%s\", ...)\n",
          parent, whitespace, format ? format : "(null)");
#endif /* DEBUG */

 /*
  * Range check input...
  */

  if (!format)
    return (NULL);

 /*
  * Create the node and set the text value...
  */

  if ((node = mxml_new(parent, MXML_TEXT)) != NULL)
  {
    va_start(ap, format);

    node->value.text.whitespace = whitespace;
    node->value.text.string     = _mxml_vstrdupf(format, ap);

    va_end(ap);
  }

  return (node);
}


/*
 * 'mxmlRemove()' - Remove a node from its parent.
 *
 * Does not free memory used by the node - use mxmlDelete() for that.
 * This function does nothing if the node has no parent.
 */

void
mxmlRemove(mxml_node_t *node)		/* I - Node to remove */
{
#ifdef DEBUG
  fprintf(stderr, "mxmlRemove(node=%p)\n", node);
#endif /* DEBUG */

 /*
  * Range check input...
  */

  if (!node || !node->parent)
    return;

 /*
  * Remove from parent...
  */

#if DEBUG > 1
  fprintf(stderr, "    BEFORE: node->parent=%p\n", node->parent);
  if (node->parent)
  {
    fprintf(stderr, "    BEFORE: node->parent->child=%p\n", node->parent->child);
    fprintf(stderr, "    BEFORE: node->parent->last_child=%p\n", node->parent->last_child);
  }
  fprintf(stderr, "    BEFORE: node->child=%p\n", node->child);
  fprintf(stderr, "    BEFORE: node->last_child=%p\n", node->last_child);
  fprintf(stderr, "    BEFORE: node->prev=%p\n", node->prev);
  fprintf(stderr, "    BEFORE: node->next=%p\n", node->next);
#endif /* DEBUG > 1 */

  if (node->prev)
    node->prev->next = node->next;
  else
    node->parent->child = node->next;

  if (node->next)
    node->next->prev = node->prev;
  else
    node->parent->last_child = node->prev;

  node->parent = NULL;
  node->prev   = NULL;
  node->next   = NULL;

#if DEBUG > 1
  fprintf(stderr, "    AFTER: node->parent=%p\n", node->parent);
  if (node->parent)
  {
    fprintf(stderr, "    AFTER: node->parent->child=%p\n", node->parent->child);
    fprintf(stderr, "    AFTER: node->parent->last_child=%p\n", node->parent->last_child);
  }
  fprintf(stderr, "    AFTER: node->child=%p\n", node->child);
  fprintf(stderr, "    AFTER: node->last_child=%p\n", node->last_child);
  fprintf(stderr, "    AFTER: node->prev=%p\n", node->prev);
  fprintf(stderr, "    AFTER: node->next=%p\n", node->next);
#endif /* DEBUG > 1 */
}


/*
 * 'mxmlNewXML()' - Create a new XML document tree.
 *
 * The "version" argument specifies the version number to put in the
 * ?xml element node. If NULL, version 1.0 is assumed.
 *
 * @since Mini-XML 2.3@
 */

mxml_node_t *				/* O - New ?xml node */
mxmlNewXML(const char *version)		/* I - Version number to use */
{
  char	element[1024];			/* Element text */


  snprintf(element, sizeof(element), "?xml version=\"%s\" encoding=\"utf-8\"?",
           version ? version : "1.0");

  return (mxmlNewElement(NULL, element));
}


/*
 * 'mxmlRelease()' - Release a node.
 *
 * When the reference count reaches zero, the node (and any children)
 * is deleted via mxmlDelete().
 *
 * @since Mini-XML 2.3@
 */

int					/* O - New reference count */
mxmlRelease(mxml_node_t *node)		/* I - Node */
{
  if (node)
  {
    if ((-- node->ref_count) <= 0)
    {
      mxmlDelete(node);
      return (0);
    }
    else
      return (node->ref_count);
  }
  else
    return (-1);
}


/*
 * 'mxmlRetain()' - Retain a node.
 *
 * @since Mini-XML 2.3@
 */

int					/* O - New reference count */
mxmlRetain(mxml_node_t *node)		/* I - Node */
{
  if (node)
    return (++ node->ref_count);
  else
    return (-1);
}


/*
 * 'mxml_new()' - Create a new node.
 */

static mxml_node_t *			/* O - New node */
mxml_new(mxml_node_t *parent,		/* I - Parent node */
         mxml_type_t type)		/* I - Node type */
{
  mxml_node_t	*node;			/* New node */


#if DEBUG > 1
  fprintf(stderr, "mxml_new(parent=%p, type=%d)\n", parent, type);
#endif /* DEBUG > 1 */

 /*
  * Allocate memory for the node...
  */

  if ((node = calloc(1, sizeof(mxml_node_t))) == NULL)
  {
#if DEBUG > 1
    fputs("    returning NULL\n", stderr);
#endif /* DEBUG > 1 */

    return (NULL);
  }

#if DEBUG > 1
  fprintf(stderr, "    returning %p\n", node);
#endif /* DEBUG > 1 */

 /*
  * Set the node type...
  */

  node->type      = type;
  node->ref_count = 1;

 /*
  * Add to the parent if present...
  */

  if (parent)
    mxmlAdd(parent, MXML_ADD_AFTER, MXML_ADD_TO_PARENT, node);

 /*
  * Return the new node...
  */

  return (node);
}

